src/model.rs

```diff
@@ -43,14 +43,27 @@ impl Into<String> for ModelType {
 impl From<ModelComponent> for cyclonedx::Component {
     // TODO: Error
     fn from(model_component: ModelComponent) -> Self {
-        cyclonedx::ComponentBuilder::default()
+        let mut builder = cyclonedx::ComponentBuilder::default();
+        let mut builder = builder
             .type_(model_component.r#type)
-            .name(model_component.name)
+            .name(model_component.name.clone())
             .bom_ref(model_component.r#ref)
-            .version(model_component.version)
-            .description(model_component.description)
-            .build()
-            .unwrap()
+            .description(model_component.description);
+
+        if !model_component.version.is_empty() {
+            builder = builder.version(model_component.version.clone());
+        }
+
+        if !model_component.name.is_empty() && !model_component.version.is_empty() {
+            let purl: String = format!(
+                "pkg:generic/{}@{}",
```

Comment:
```suggestion
                "pkg:nix/{}@{}",
```

Should it be `nix` scope instead?

Reply:
Hej! We have discussed this internally, let me give you a summary of our decision here.

First on the potential `nix` purl. Whether it is part of the spec is definitely important to us, but more important is its inclusion in CVE databases (which we consider the main usecase of the purls). Such databases typically only have a subset of purls, so even if `nix` was included, these databases are likely to lag behind. In such cases we would rather be compatible with the databases (which do typically contain the generic purl).

Keep in mind that the cycloneDX spec only allows for a single purl to be associated with a component.

Regarding using upstream purls, this is definitely something we want to do, but is planned for later.

Reply:
Ah gotcha, thanks for clarifying!

Reply:
Some more brainstorming on a "nix" PURL type: I think the type should be `nixpkgs` instead of `nix`, the reason being that Nixpkgs has an architecture and metadata that only it enforces and that are valuable assumptions to make PURL meaningful. Also it should probably namespaced by nixpkgs revision/tag. The rest of the path would be the attribute path. In the end we'd have for instance `pkg:nixpkgs/23.11/gcc`, `pkg:nixpkgs/23.11/python3Packages.setuptools`. I see that the arch is often set as a query param of the URL, so I would suggest to keep it as is.

In the meantime I think `generic` has more chances to hit on CVE databases as @ErinvanderVeen explained.

Reply:
> I think the type should be nixpkgs instead of nix, the reason being that Nixpkgs has an architecture and metadata that only it enforces and that are valuable assumptions to make PURL meaningful.

Yes, I agree that `nixpkgs` have that practice on filling in metadata and I don't think that someone often uses `builtins.derivaton` directly outside of `nixpkgs`, because it's more convenient to use `nixpkgs.mkDerivation` to create a package. But the resulting package doesn't become a part of `nixpkgs` after that fact, right?

I guess my question is how about standalone packages' implementations as Nix Flakes that don't belong to `nixpkgs` set but do provide third-party software, for example like this [prettier flake](https://github.com/olebedev/prettier.nix/blob/main/flake.nix)? Should we target such packages as well?

To me it sounds that `nixpkgs` is a _subtype_ of `nix` _supertype_, but I can be wrong here :D Please let me know what are your thoughts.

Reply:
Upon a closer look, I came across [the spec for deb](https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst#deb):

> * There is no default package repository: this should be implied either from the `distro` qualifiers key or using a base url as a `repository_url` qualifiers key.

So indeed we could go with `nix` and specify alternative repos using `repository_url`.

Reply:
There is existing discussion around a `nix` purl type at https://discourse.nixos.org/t/package-urls-purl-for-nix-packages/27934 . It would be really good if the attribute path could be in there where possible, not just a pname and drv/output hash ;).

While I think it would be a great idea to flesh that purl type out further, note that for the purpose of advisory (CVE/OSV/..) mapping, when we design a nix purl we should also immediately make a 'mapping' from the nix purl to the purl types that are more likely to be found in those advisory databases, such as `generic`, `github`, `pypi`, `maven` etc. It would be really helpful if the SBOM generator would also provide that mapping directly in the SBOM - I know it's tempting to consider that out-of-scope but it would make the result much more useful :) .

CycloneDX only has a single `purl` field, but it looks like you could use a `pedigree.ancestor` to store the more 'generic' purl type(s). That also nicely allows making the distinction between 'the project itself' and 'the nix packaging of the project' where needed.